cpp/src/arrow/buffer_test.cc

```diff
@@ -173,11 +173,11 @@ TEST(TestMutableBuffer, Wrap) {
 TEST(TestBuffer, FromStringRvalue) {
   std::string expected = "input data";
 
-  std::shared_ptr<Buffer> buffer;
-  {
-    std::string data_str = "input data";
-    buffer = Buffer::FromString(std::move(data_str));
-  }
+  std::string data_str = "input data";
+  std::shared_ptr<Buffer> buffer = Buffer::FromString(std::move(data_str));
+
+  // data_str is moved
+  ASSERT_EQ(0, data_str.size());
```

Comment:
Instead, I would check that the pointer is the same.

Reply:
That will break in this case due to the short string optimization

Reply:
Ahah.

Reply:
But I think it's the right way to go: construct `expected` such that `expected.size() > sizeof(std::string)`, then store the data pointer before the move. It should be identical to `buffer->data()`